test/provider/database.unittest.ts

```diff
@@ -0,0 +1,210 @@
+        });
+    });
+
+    describe('Upsert', () => {
+        it('Adds a new record', async () => {
+            const note = generateNotification();
+
+            await collection.upsert(note);
+            expect(await collection.getAll()).toEqual([note]);
+        });
+
+        it('Adds many records from array', async () => {
+            const notes = [generateNotification(), generateNotification(), generateNotification()];
+            await collection.upsert(notes);
+
+            const results = await collection.getAll();
+            expect(results).toEqual(notes);
+        });
+
+        it('Updates an existing record', async () => {
+            const [note1, note2] = [generateNotification({id: 'updateable'}), generateNotification({id: 'updateable', title: 'title'})];
+
+            await collection.upsert(note1);
+            await collection.upsert(note2);
+
+            const results = await collection.getAll();
+            expect(results).toEqual([note2]);
+        });
+
+        it('Updates many existing records from array', async () => {
+            const expectedTitle = 'updated title';
+            const notes = [generateNotification({id: 'updateable'}), generateNotification({id: 'updateable2'}), generateNotification()];
+            await collection.upsert(notes);
+
+            const updatedNotes = [generateNotification({id: 'updateable', title: expectedTitle}),
+                generateNotification({id: 'updateable2', title: expectedTitle})];
+            await collection.upsert(updatedNotes);
+
+            const filteredNotes = (await collection.getAll()).filter(note => note.notification.title === expectedTitle);
+            expect(filteredNotes.length).toEqual(updatedNotes.length);
+        });
+
+        it('Updates existing records and adds new records from array', async () => {
+            const expectedTitle = 'updated title';
+            await collection.upsert(generateNotification({id: 'updateable'}));
+
+            const newAndUpdatedNotes = [generateNotification(), generateNotification({id: 'updateable', title: expectedTitle})];
+            await collection.upsert(newAndUpdatedNotes);
+
+            const results = await collection.getAll();
+
+            expect(results).toEqual(newAndUpdatedNotes);
+        });
+    });
+
+    describe('Delete', () => {
+        it('Deletes a single record', async () => {
+            const note = generateNotification();
+            await collection.upsert(note);
+            await collection.delete(note.id);
+            expect(await collection.getAll()).toEqual([]);
+        });
+
+        it('Deletes many records from array', async () => {
+            const notes = [generateNotification(), generateNotification(), generateNotification()];
+            await collection.upsert(notes);
+            await collection.delete(notes.map(note => note.id));
+
+            expect(await collection.getAll()).toEqual([]);
+        });
```

Comment:
Suggest add a 'Deletes single record from an array' test

Reply:
As discussed, I have added more notifications to the `Deletes a single record` test